### PR TITLE
Store program admin access ACLs on stored files

### PR DESCRIPTION
**Pull request by bion**
This resolves a bug with program admin access to applicant-uploaded files that occurs when a file upload question is reused between programs. The old auth logic relied on the program ID embedded in the file key to check if the admin has access to the program identified by the ID. This broke when applicants applied to another program that reused the question since the admins of the second program may not have read access to the first.

This PR introduces ACLs for the `StoredFile` resource. When an applicant submits an application, read permission for the program they applied to are stored in the ACLs for each file they included in their application. When an admin attempts to view a file, CiviForm checks that they are an admin for at least one of the programs in the file's ACLs.

The original route is retained (and will be indefinitely) to support previously exported file URLs. A new route is added though that doesn't include the program ID in the URL path.

- [x] store ACLs on StoredFile
- [x] when an application is submitted, update all of its stored files' ACLs to include the program name
- [x] for program admins viewing files, check the stored file's ACLs for permission
- [x] original route and auth logic kept intact to support previously issued file URLs
- [x] documentation
- [x] unit tests
- [ ] ~browser test~ unable to write a browser test for this with current browser testing abilities because for the program admin role in test mode we create an account with program admin ownership of all programs

### Issue(s)
Fixes https://github.com/seattle-uat/civiform/issues/2331, https://github.com/seattle-uat/civiform/issues/2547

### CODE REVIEW

File: server/app/views/admin/programs/ProgramApplicationView.java [hunk 1]
```diff
@@ -113,14 +109,13 @@ private Tag renderApplicationBlock(long programId, Block block, Collection<Answe
...
-  private Tag renderAnswer(long programId, AnswerData answerData) {
+  private Tag renderAnswer(AnswerData answerData) {
     LocalDate date =
         Instant.ofEpochMilli(answerData.timestamp()).atZone(ZoneId.systemDefault()).toLocalDate();
     Tag answerContent;
     if (answerData.fileKey().isPresent()) {
       String encodedFileKey = URLEncoder.encode(answerData.fileKey().get(), StandardCharsets.UTF_8);
-      String fileLink =
-          controllers.routes.FileController.adminShow(programId, encodedFileKey).url();
+      String fileLink = controllers.routes.FileController.acledAdminShow(encodedFileKey).url();
```
Comment (kmcnellis) [inline server/app/views/admin/programs/ProgramApplicationView.java:118]: How will these work for legacy files?

File: server/test/controllers/FileControllerTest.java [hunk 2]
```diff
@@ -58,28 +57,18 @@ public void show_TIManagedApplicant_redirects() {
...
-  }
-
   @Test
   public void adminShow_differentProgram_returnsUnauthorizedResult() {
     Program programOne = ProgramBuilder.newDraftProgram("one").build();
     Program programTwo = ProgramBuilder.newDraftProgram("two").build();
     createProgramAdminWithMockedProfile(programOne);
     String fileKey = fakeFileKey(1L, programTwo.id);
-    Request request = fakeRequest().build();
+    createStoredFileWithProgramAccess(fileKey, programTwo);
```
Comment (kmcnellis) [inline server/test/controllers/FileControllerTest.java:70]: Are we still testing files that don't have ACLS (legacy)?

File: server/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java [hunk 3]
```diff
@@ -73,6 +73,47 @@ public void getProgramTitle_returnsProgramTitleForPreferredLocale() {
     assertThat(subject.getProgramTitle()).isEqualTo("Mein Programm");
   }
 
+  @Test
+  public void getStoredFileKeys_includesAnsweredFileQuestions() {
+    QuestionDefinition fileQuestionDefinition =
```
Comment (kmcnellis) [inline server/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java:78]: Add one for includes legacy files

File: server/test/services/export/JsonExporterTest.java [hunk 4]
```diff
@@ -44,9 +44,7 @@ public void testAllQuestionTypesWithoutEnumerators() throws Exception {
         2, ".applicant_favorite_color.text", "Some Value \" containing ,,, special characters");
     resultAsserter.assertValueAtPath(2, ".applicant_monthly_income.currency_dollars", 1234.56);
     resultAsserter.assertValueAtPath(
-        2,
-        ".applicant_file.file_key",
-        "http://localhost:9000/admin/programs/" + fakeProgram.id + "/files/my-file-key");
+        2, ".applicant_file.file_key", "http://localhost:9000/admin/applicant-files/my-file-key");
```
Comment (kmcnellis) [inline server/test/services/export/JsonExporterTest.java:47]: This should include a legacy file path

**Review by kmcnellis — DISMISSED** [review]

**Comment by shanemc-goog** [comment]
Can you please flesh out the description with more beyond a checklist?   A small stand in for there being no design doc around file storage ACLs.

Neither issue or the PR describes the actual code issue nor how the solution addresses it, and in what manner it does it.

File: server/app/views/admin/programs/ProgramApplicationView.java (hunk 1 above)
Comment (shanemc-goog) [reply server/app/views/admin/programs/ProgramApplicationView.java:118]: +1

File: server/app/services/applicant/ApplicantServiceImpl.java [hunk 5]
```diff
@@ -234,19 +240,60 @@ private CompletionStage<Application> submitApplication(
...
+              return updateStoredFileAclsForSubmit(applicantId, programId)
+                  .thenApplyAsync((ignoreVoid) -> application, httpExecutionContext.current());
+            },
+            httpExecutionContext.current());
+  }
+
+  /**
+   * When an application is submitted, we store the name of its program in the ACLs for each file in
+   * the application. This is so that when a program admin goes to view it we can authorize them,
+   * both for the original program the file was uploaded for as well as programs that the applicant
```
Comment (shanemc-goog) [inline server/app/services/applicant/ApplicantServiceImpl.java:268]: I'd remove the original part and speak about how this works in general.  The "original" part is speaking towards legacy support that is otherwise not present or a concern in this class.

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java [hunk 6]
```diff
@@ -71,6 +71,20 @@ public String getProgramTitle() {
...
   }
 
+  @Override
+  public ImmutableList<String> getStoredFileKeys() {
+    return getAllActiveBlocks().stream()
+        .filter(Block::isFileUpload)
+        .flatMap(
+            block ->
+                block.getQuestions().stream()
+                    .filter(question -> question.isAnswered() && question.isFileUploadQuestion()))
```
Comment (shanemc-goog) [inline server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:81]: all of these can be their own operations in the parent stream.

.flatMap(Block::getQuestions)
.filter(Question::isAnswered)
.filter(Question::isFileUploadQuestion)

File: server/app/auth/StoredFileAcls.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    return ImmutableSet.copyOf(programReadAcls);
+  }
+
+  public StoredFileAcls addProgramToReaders(ProgramDefinition programDefinition) {
+    this.programReadAcls.add(programDefinition.adminName());
+    return this;
+  }
+
+  public boolean hasProgramReadPermission(Account account) {
+    for (String administeredProgramName : account.getAdministeredProgramNames()) {
```
Comment (shanemc-goog) [inline server/app/auth/StoredFileAcls.java:42]: this method can be rewritten as

```return account.getAdministeredProgramNames().stream().anyMatch(programReadAcls::contains);```

File: server/test/auth/StoredFileAclsTest.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+public class StoredFileAclsTest extends ResetPostgres {
+
+  private ProgramDefinition programOne =
+      ProgramBuilder.newDraftProgram("program-one").buildDefinition();
+  private ProgramDefinition programTwo =
+      ProgramBuilder.newDraftProgram("program-two").buildDefinition();
+
+  @Test
+  public void hasProgramReadPermission_hasPermission() {
+    var acls = new StoredFileAcls().addProgramToReaders(programOne);
```
Comment (shanemc-goog) [inline server/test/auth/StoredFileAclsTest.java:20]: Add a test for the empty acl case since all existing files will be in that state.

File: server/test/controllers/FileControllerTest.java [hunk 9]
```diff
@@ -58,28 +57,18 @@ public void show_TIManagedApplicant_redirects() {
   public void show_redirects() {
     Applicant applicant = createApplicantWithMockedProfile();
     String fileKey = fakeFileKey(applicant.id, 1L);
-    Request request = fakeRequest().build();
     Result result = controller.show(request, applicant.id, fileKey).toCompletableFuture().join();
     assertThat(result.status()).isEqualTo(SEE_OTHER);
   }
 
-  @Test
-  public void adminShow_invalidProgram_returnsNotFound() {
```
Comment (shanemc-goog) [inline server/test/controllers/FileControllerTest.java:67]: Why remove this? The handler is still valid.

**Review by shanemc-goog — CHANGES_REQUESTED** [review]

File: server/app/services/applicant/ApplicantServiceImpl.java (hunk 5 above)
Comment (bion) [reply server/app/services/applicant/ApplicantServiceImpl.java:268]: > The "original" part is speaking towards legacy support that is otherwise not present or a concern in this class.

It's not speaking toward legacy support. "original" is talking about the first program the file was uploaded for. The entire paragraph his discussing new behavior.

Comment (shanemc-goog) [reply server/app/services/applicant/ApplicantServiceImpl.java:268]: ah ok. Is that actually relevant for the new ACL model though? I don't think the order of programs the applicant applied too (and this is updated for) is relevant.

File: server/conf/evolutions/default/39.sql [hunk 10]
```diff
@@ -0,0 +1,9 @@
+# --- Add acls to files
+
+# --- !Ups
+CREATE UNIQUE INDEX idx_file_names ON files (name);
```
[the remark this answers is not in the document]
Comment (bion) [reply server/conf/evolutions/default/39.sql:4]: done

File: server/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java (hunk 3 above)
Comment (bion) [reply server/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java:78]: I'm not sure what that means in the context of this class/test. This class doesn't have logic that differentiates between the two, nor does it deal directly with the `StoredFile` class.

File: server/app/services/applicant/ApplicantServiceImpl.java (hunk 5 above)
Comment (bion) [reply server/app/services/applicant/ApplicantServiceImpl.java:268]: I want to make sure it's clear why we have extra logic here for stored files vs all the other data that doesn't require this kind of ACLing behavior. Do you think the comment is adequate for that if I omit the second sentence?

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java (hunk 6 above)
Comment (bion) [reply server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:81]: done. though i was disappointed in java to learn that i couldn't do `.flatMap(Block::getQuestions)` because flatMap doesn't know how to turn an `ImmutableList` into a stream so i had to do `.flatMap(block -> block.getQuestions().stream())`

File: server/app/models/StoredFile.java [hunk 11]
```diff
@@ -11,20 +15,42 @@
...
     this.name = name;
+    return this;
   }
 
   public Optional<String> getOriginalFileName() {
     return Optional.ofNullable(originalFileName);
   }
 
-  public void setOriginalFileName(String originalFileName) {
+  public StoredFile setOriginalFileName(String originalFileName) {
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/models/StoredFile.java:51]: setter is returning the stored file (the object being modified), not the filename. returning `this` is a common pattern for setters throughout the code base that allows us to chain method calls e.g.

```java
model
  .setPropA(1)
  .setPropB(2)
  .setPropC(3)
``

File: server/app/views/admin/programs/ProgramApplicationView.java (hunk 1 above)
Comment (bion) [reply server/app/views/admin/programs/ProgramApplicationView.java]: Yeah, thank you. Coming back to this and realizing we need to migrate the old file data to have ACLs before we can begin issuing the new URLs

File: server/app/controllers/FileController.java [hunk 12]
```diff
@@ -66,6 +73,44 @@ public CompletionStage<Result> show(Request request, long applicantId, String fi
...
   public Result adminShow(Request request, long programId, String fileKey) {
+    return adminShowInternal(request, Optional.of(programId), fileKey);
+  }
+
+  @Secure(authorizers = Authorizers.Labels.ANY_ADMIN)
+  public Result acledAdminShow(Request request, String fileKey) {
+    return adminShowInternal(request, /* maybeProgramId= */ Optional.empty(), fileKey);
+  }
+
+  public Result adminShowInternal(Request request, Optional<Long> maybeProgramId, String fileKey) {
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/controllers/FileController.java:84]: done

File: server/app/controllers/FileController.java [hunk 13]
```diff
@@ -66,6 +73,44 @@ public CompletionStage<Result> show(Request request, long applicantId, String fi
 
   @Secure(authorizers = Authorizers.Labels.ANY_ADMIN)
   public Result adminShow(Request request, long programId, String fileKey) {
+    return adminShowInternal(request, Optional.of(programId), fileKey);
+  }
+
+  @Secure(authorizers = Authorizers.Labels.ANY_ADMIN)
+  public Result acledAdminShow(Request request, String fileKey) {
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/controllers/FileController.java:80]: done

File: server/app/controllers/FileController.java [hunk 14]
```diff
@@ -66,6 +73,44 @@ public CompletionStage<Result> show(Request request, long applicantId, String fi
...
+    String decodedFileKey = URLDecoder.decode(fileKey, StandardCharsets.UTF_8);
+    return redirect(storageClient.getPresignedUrlString(decodedFileKey));
+  }
+
+  /**
+   * The old auth logic assumes that the admin viewing the file should be an admin for the original
+   * program the file was uploaded for. This is not the case, since file upload questions can be
+   * shared between programs.
+   */
+  private Result legacyAdminShow(Request request, long programId, String fileKey) {
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/controllers/FileController.java:113]: done

[the remark this answers is not in the document]
Comment (bion) [reply server/app/controllers/FileController.java:113]: I didn't change the file key format since it still serves it's purpose of creating a unique and readable file key for each uploaded file. The program ID embedded in it is useful for that, but we don't need to use it for auth logic anymore.

[the remark this answers is not in the document]
Comment (bion) [reply server/app/controllers/FileController.java:133]: done

File: server/test/controllers/FileControllerTest.java (hunk 9 above)
Comment (bion) [reply server/test/controllers/FileControllerTest.java]: done

File: server/conf/routes [hunk 15]
```diff
@@ -92,6 +92,7 @@ POST    /admin/apiKeys                              controllers.admin.AdminApiKe
 POST    /admin/apiKeys/:apiKeyId/retire             controllers.admin.AdminApiKeysController.retire(request: Request, apiKeyId: Long)
 
 # Controller for admins and applicants to access applicants' uploaded files
+GET     /admin/applicant-files/:fileKey             controllers.FileController.acledAdminShow(request: Request, fileKey: String)
```
[the remark this answers is not in the document]
Comment (bion) [reply server/conf/routes:95]: Tried to add more docs in relevant places.

The end result will be both routes remain (to support previously exported URLs) but will use the same auth logic after the data is migrated.

File: server/app/controllers/FileController.java [hunk 16]
```diff
@@ -66,6 +73,44 @@ public CompletionStage<Result> show(Request request, long applicantId, String fi
...
+          .map(programId -> legacyAdminShow(request, programId, fileKey))
+          .orElse(unauthorized());
+    }
+
+    String decodedFileKey = URLDecoder.decode(fileKey, StandardCharsets.UTF_8);
+    return redirect(storageClient.getPresignedUrlString(decodedFileKey));
+  }
+
+  /**
+   * The old auth logic assumes that the admin viewing the file should be an admin for the original
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/controllers/FileController.java:127]: Added a TODO

File: server/test/auth/StoredFileAclsTest.java (hunk 8 above)
Comment (bion) [reply server/test/auth/StoredFileAclsTest.java:20]: done

File: server/test/controllers/FileControllerTest.java (hunk 9 above)
Comment (bion) [reply server/test/controllers/FileControllerTest.java]: added it back

File: server/app/controllers/FileController.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (shanemc-goog) [reply server/app/controllers/FileController.java:133]: Ah ok, so likely we just need to give each file a unique (db sequence/id) value for uniqueness and not rely on the program construction that it was created under for that pseudo uniqueness.

File: server/app/controllers/FileController.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (shanemc-goog) [reply server/app/controllers/FileController.java:127]: Can we remove legacy support after that? Include it if so. EG the TODO is to add that and remove this legacy code.

File: server/app/services/applicant/ApplicantServiceImpl.java (hunk 5 above)
Comment (shanemc-goog) [reply server/app/services/applicant/ApplicantServiceImpl.java:268]: Yeah, I'd suggest just keeping the first sentence as it's directly pertinent to this code.  

When someone reads the code base, they don't know all the history, so they're going to see this as _the_ acl mechanism.  The legacy concerns are then documented and isolated to where they are relevant; and I believe can be removed in the near future via the TODOs you added.

File: server/app/controllers/FileController.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (bion) [reply server/app/controllers/FileController.java:127]: We probably shouldn't since there have been thousands of applications exported that all have URLs with the old URL path.

File: server/app/controllers/FileController.java [hunk 17]
```diff
@@ -64,8 +71,66 @@ public CompletionStage<Result> show(Request request, long applicantId, String fi
...
+   * program the file was uploaded for. This is not the case, since file upload questions can be
+   * shared between programs.
+   */
+  // TODO(https://github.com/seattle-uat/civiform/issues/2709): add ACLs to previously uploaded
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/controllers/FileController.java:131]: yeah

File: server/app/services/applicant/ApplicantServiceImpl.java (hunk 5 above)
Comment (bion) [reply server/app/services/applicant/ApplicantServiceImpl.java:268]: sounds good, done

File: server/app/services/cloud/FileNameFormatter.java [hunk 18]
```diff
@@ -10,37 +8,14 @@
...
-        params.applicantId(), params.programId(), params.block().getId());
-  }
-
-  /**
-   * This returns a prefixed filename used in the dev file upload view, which will be persisted in
-   * the DB.
-   */
-  public static String formatDevUploadFilename() {
-    return "dev/${filename}";
+        "applicant-%d/program-%d/block-%s/${filename}", applicantId, programId, blockId);
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/services/cloud/FileNameFormatter.java:19]: yep, intentional

File: server/conf/routes (hunk 15 above)
[the remark this answers is not in the document]
Comment (bion) [reply server/conf/routes:95]: yes, for previously exported applications. unless at some point seattle tells us they're ok with reexporting all of them

**Review by shanemc-goog — APPROVED** [review]
